### PR TITLE
(#1565) Improve generic variance for `org.cactoos.iterator`

### PR DESCRIPTION
**Pull request by DmitryBarskov**
For #1565:

Please also see #1611.

Add covariant constructor params for

- `Reversed`
- `Cycled`
- `Repeated`
- `Endless`
- `Immutable`

Generify `RangeOf(T, T, Func<T, T>)`

### CODE REVIEW

**Comment by victornoel** [comment]
@0crat status

**Comment by victornoel** [comment]
@0crat status

**Comment by victornoel** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/yegor256/cactoos/issues/1612#issuecomment-860062879))

@victornoel This is what I know about this job in C63314D6Z, as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#1612](https://github.com/yegor256/cactoos/issues/1612) is not in scope
  * Job [footprint](https://www.0crat.com/footprint/C63314D6Z?q=%7Bjob%3A%27gh%3Ayegor256%2Fcactoos%231612%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

**Comment by victornoel** [comment]
@0crat in

File: src/main/java/org/cactoos/iterator/Endless.java [hunk 1]
```diff
@@ -56,15 +56,15 @@ public Endless(final T element) {
...
+    public Endless(final Scalar<? extends T> scalar) {
         this(new Unchecked<>(scalar));
     }
 
     /**
      * Ctor.
      * @param scalar Scalar to repeat
      */
-    public Endless(final Unchecked<T> scalar) {
+    public Endless(final Unchecked<? extends T> scalar) {
```
Comment (andreoss) [inline src/main/java/org/cactoos/iterator/Endless.java:67]: @DmitryBarskov This ctor should be either private or removed, there should be no concrete parameters in ctors

File: src/main/java/org/cactoos/iterator/Repeated.java [hunk 2]
```diff
@@ -62,16 +62,16 @@ public Repeated(final int max, final T element) {
...
+        this(max, new Unchecked<>(scalar));
     }
 
     /**
      * Ctor.
      * @param max How many times to repeat
      * @param scalar Scalar to repeat
      */
-    private Repeated(final int max, final Unchecked<T> scalar) {
+    private Repeated(final int max, final Unchecked<? extends T> scalar) {
```
Comment (andreoss) [inline src/main/java/org/cactoos/iterator/Repeated.java:74]: @DmitryBarskov I suggest removing this ctor, and doing `this.elm = new Unchecked<>(...)` in ctor itself

**Review by andreoss — CHANGES_REQUESTED** [review]

**Comment by DmitryBarskov** [comment]
@andreoss do you have any suggestions how to improve coverage for `Repeated`?

**Comment by victornoel** [comment]
@DmitryBarskov I think the best to have the whole CI passing would be to simply increase test coverage independently of what was changed in this PR. You can also safely ignore the error since it's "just" -0.01%" :)

**Comment by victornoel** [comment]
@DmitryBarskov did you validate that along with the changes done in #1611 there are no more need for change sin the iterator package?

**Comment by DmitryBarskov** [comment]
> @DmitryBarskov did you validate that along with the changes done in #1611 there are no more need for change sin the iterator package?

@victornoel yes, no changes needed (related to the ticket)

**Comment by victornoel** [comment]
@rultor merge

**Comment by 0crat** [comment]
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @andreoss[/z](https://www.0crat.com/u/andreoss), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by sereshqua** [comment]
@andreoss please make sure you will find at least 3 issues during next CR, thanks

**Comment by sereshqua** [comment]
@0crat quality acceptable

**Pull request merged**